### PR TITLE
Add example showing how to use SpecializedMeshPipeline

### PR DESCRIPTION
**Pull request by IceSentry**
# Objective

- A lot of mid-level rendering apis are hard to figure out because they don't have any examples
- SpecializedMeshPipeline can be really useful in some cases when you want more flexibility than a Material without having to go to low level apis.

## Solution

- Add an example showing how to make a custom `SpecializedMeshPipeline`.

## Testing

- Did you test these changes? If so, how?
- Are there any parts that need more testing?
- How can other people (reviewers) test your changes? Is there anything specific they need to know?
- If relevant, what platforms did you test these changes on, and are there any important ones you can't test?

---

## Showcase

The examples just spawns 3 triangles in a triangle pattern.

image

### CODE REVIEW

File: assets/shaders/specialized_mesh_pipeline.wgsl [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    var world_from_local = mesh_functions::get_world_from_local(vertex.instance_index);
+    out.world_position = mesh_functions::mesh_position_local_to_world(world_from_local, vec4(vertex.position, 1.0));
+    out.clip_position = position_world_to_clip(out.world_position.xyz);
+    out.color = vertex.color.rgb;
+    return out;
+}
+
+@fragment
+fn fragment(in: VertexOutput) -> @location(0) vec4<f32> {
+    return vec4(in.color, 1.0);
```
Comment (tychedelia) [inline assets/shaders/specialized_mesh_pipeline.wgsl:30]: nit: could this use world position? even if a bit silly like dividing the color by the position or smth

File: examples/shader/specialized_mesh_pipeline.rs [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+            depth_stencil: Some(DepthStencilState {
+                format: CORE_3D_DEPTH_FORMAT,
+                depth_write_enabled: true,
+                depth_compare: CompareFunction::GreaterEqual,
+                stencil: default(),
+                bias: default(),
+            }),
+            // This is isn't required, but if you can support MSAA
+            // it's generally recommended to specialize your pipeline for it
+            multisample: MultisampleState {
```
Comment (tychedelia) [inline examples/shader/specialized_mesh_pipeline.rs:266]: might be making this up pretty sure this is required and will panic if it's not set correctly?

**Review by tychedelia — APPROVED** [review]
ty!! so helpful to have more of these mid-level examples

File: examples/shader/specialized_mesh_pipeline.rs (hunk 2 above)
Comment (IceSentry) [reply examples/shader/specialized_mesh_pipeline.rs:266]: If you don't use MSAA it's not required. 

Well, I should update the comment a bit, setting the sample count isn't required, but you still need to use `MultisampleState::default()`

File: assets/shaders/specialized_mesh_pipeline.wgsl (hunk 1 above)
Comment (IceSentry) [reply assets/shaders/specialized_mesh_pipeline.wgsl:30]: I'd like to keep the shader as simple as possible. I think people using that level of api are able to figure that out. I could maybe rotate the camera a bit to show that it's all already correctly placed in the world.

Comment (IceSentry) [reply assets/shaders/specialized_mesh_pipeline.wgsl:30]: I should probably add a few comments about the world position and instance_index though.

**Review by viridia — APPROVED** [review]

**Comment by NotAFile** [comment]
This seems very helpful, I did not know this functionality existed. However, with this example I'm still slightly confused what the use cases and limitations are and when to use it.

**Comment by IceSentry** [comment]
> I'm still slightly confused what the use cases and limitations are and when to use it

It's a bit hard to explain if, for example, using the material apis was enough for your use case. This api is useful when you want to have more control over how a mesh is rendered than is possible with the material api.

**Comment by NotAFile** [comment]
Ah yep, that seems useful information to add to the example! They're often a bit short on information as to *why* to do things, I find.

**Comment by alice-i-cecile** [comment]
This has subtle merge conflicts with https://github.com/bevyengine/bevy/pull/14273 that need to be fixed manually <3

**Pull request merged**